stable_baselines3/common/pytree_dataclass.py

```diff
@@ -1,21 +1,282 @@
+    """Metaclass to register dataclasses as PyTrees.
+
+    Usage:
+      class MyDataclass(metaclass=_DataclassPyTreeMeta):
+        ...
+    """
+
+    # We need to have this `currently_registering` variable because, in the course of making a DataClass with __slots__,
+    # another class is created. So this will be called *twice* for every dataclass we annotate with this metaclass.
+    currently_registering: ClassVar[Optional[type]] = None
+
+    def __new__(mcs, name, bases, namespace, slots=True, **kwargs):
+        # First: create the class in the normal way.
+        cls = super().__new__(mcs, name, bases, namespace)
+
+        if dataclasses.is_dataclass(cls):
+            # If the class we're registering is already a Dataclass, it means it is a descendant of FrozenPyTreeDataclass or
+            # MutablePyTreeDataclass.
+            # This includes the children which are created when we create a dataclass with __slots__.
+
+            if mcs.currently_registering is not None:
+                # We've already created and annotated a class without __slots__, now we create the one with __slots__
+                # that will actually get returned after from the __new__ method.
+                assert mcs.currently_registering.__module__ == cls.__module__
+                assert mcs.currently_registering.__name__ == cls.__name__
+                mcs.currently_registering = None
+                return cls
+
+            else:
+                assert name not in _RESERVED_NAMES, (
+                    f"Class with name {name}: classes {_RESERVED_NAMES} don't inherit from a dataclass, so they should "
+                    "not be in this branch."
+                )
+
+                # Otherwise we just mark the current class as what we're registering.
+                if not issubclass(cls, (FrozenPyTreeDataclass, MutablePyTreeDataclass)):
+                    raise TypeError(f"Dataclass {cls} should inherit from FrozenPyTreeDataclass or MutablePyTreeDataclass")
+                mcs.currently_registering = cls
+        else:
+            mcs.currently_registering = cls
+
+        if name in _RESERVED_NAMES:
+            if not (
+                namespace["__module__"] == "stable_baselines3.common.pytree_dataclass" and namespace["__qualname__"] == name
+            ):
+                raise TypeError(f"You cannot have another class named {name} with metaclass=_PyTreeDataclassMeta")
+
+            if name == "_PyTreeDataclassBase":
+                return cls
+            frozen = kwargs.pop("frozen")
+        else:
+            if "frozen" in kwargs:
+                raise TypeError(
+                    "You should not specify frozen= for descendants of FrozenPyTreeDataclass or MutablePyTreeDataclass"
+                )
+
+            frozen = issubclass(cls, FrozenPyTreeDataclass)
+            if frozen:
+                if not (not issubclass(cls, MutablePyTreeDataclass) and issubclass(cls, FrozenPyTreeDataclass)):
```

Comment:
I weakly recommend applying Demorgan's law here and rewriting as:

```
if issubclass(cls, MutablePyTreeDataclass) or not issubclass(cls, FrozenPyTreeDataclass)):
```

IMO more easily understood, but won't argue if you disagree.